### PR TITLE
daemon, client, cmd/snap: include architecture in 'snap version'

### PR DESCRIPTION
**Pull request by chipaca**
This'll be useful when we're helping users, and stop us from assuming
amd64 all the time :-)

### CODE REVIEW

**Comment by anonymouse64** [comment]
unrelated test failures on Eoan to be inspected perhaps:
<details><summary>tests/main/desktop-portal-filechooser</summary>

```
2019-09-03 22:46:31 Error executing google:ubuntu-19.10-64:tests/main/desktop-portal-filechooser : 
-----
+ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/desktop-portal.sh
++ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/pkgdb.sh
+++ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/quiet.sh
+++ id -u test
++ TEST_UID=12345
++ USER_RUNTIME_DIR=/run/user/12345
++ DBUS_SESSION_BUS_ADDRESS=
+ echo 'Install the portals test client'
Install the portals test client
+ snap install --edge test-snapd-portal-client
test-snapd-portal-client (edge) 1.0 from James Henstridge (jamesh) installed
+ echo 'The confined application can run open files via the portal'
The confined application can run open files via the portal
+ echo from-host-system
+ chown test:test /tmp/file-to-read.txt
+ MATCH from-host-system
+ as_user test-snapd-portal-client open-file
+ '[' -z '' ']'
++ grep '^DBUS_SESSION_BUS_ADDRESS'
++ su -l -c 'XDG_RUNTIME_DIR="/run/user/12345" systemctl --user show-environment' test
+ eval DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/12345/bus
++ DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/12345/bus
+ su -l -c 'XDG_RUNTIME_DIR="/run/user/12345" DBUS_SESSION_BUS_ADDRESS="unix:path=/run/user/12345/bus" test-snapd-portal-client open-file' test
+ echo 'The confined application can write files via the portal'
The confined application can write files via the portal
+ '[' '!' -f /tmp/file-to-write.txt ']'
+ touch /tmp/file-to-write.txt
+ chown test:test /tmp/file-to-write.txt
+ as_user test-snapd-portal-client save-file from-sandbox
+ '[' -z '' ']'
++ grep '^DBUS_SESSION_BUS_ADDRESS'
++ su -l -c 'XDG_RUNTIME_DIR="/run/user/12345" systemctl --user show-environment' test
+ eval DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/12345/bus
++ DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/12345/bus
+ su -l -c 'XDG_RUNTIME_DIR="/run/user/12345" DBUS_SESSION_BUS_ADDRESS="unix:path=/run/user/12345/bus" test-snapd-portal-client save-file from-sandbox' test
+ '[' -f /tmp/file-to-write.txt ']'
+ MATCH from-sandbox
grep error: pattern not found, got:
-----
travis_time:end:84c369af:start=1567550786232347055,finish=1567550791824273295,duration=5591926240
```

</details>

<details><summary>tests/main/interfaces-timeserver-control</summary>

```
2019-09-03 22:33:54 Error executing google:ubuntu-19.10-64:tests/main/interfaces-timeserver-control : 
-----
+ echo 'The interface is disconnected by default'
The interface is disconnected by default
+ MATCH -- '- +test-snapd-timedate-control-consumer:timeserver-control'
+ snap interfaces -i timeserver-control

'snap interfaces' is deprecated; use 'snap connections'.
+ echo 'When the interface is connected'
When the interface is connected
+ snap connect test-snapd-timedate-control-consumer:timeserver-control
+ grep -oP '(Network time on|systemd-timesyncd.service active): \K(.*)'
+ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ test-snapd-timedate-control-consumer.timedatectl-timeserver set-ntp yes
++ seq 10
+ for _ in $(seq 10)
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
+ sleep 1
+ for _ in $(seq 10)
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
+ sleep 1
+ for _ in $(seq 10)
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
+ sleep 1
+ for _ in $(seq 10)
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
+ sleep 1
+ for _ in $(seq 10)
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
+ sleep 1
+ for _ in $(seq 10)
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
+ sleep 1
+ for _ in $(seq 10)
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
+ sleep 1
+ for _ in $(seq 10)
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
+ sleep 1
+ for _ in $(seq 10)
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
+ sleep 1
+ for _ in $(seq 10)
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
+ sleep 1
++ get_ntp_status
++ grep -oP '(Network time on|System clock synchronized): \K(.*)'
++ test-snapd-timedate-control-consumer.timedatectl-timeserver status
+ '[' no = yes ']'
-----
travis_time:end:cd92ac84:start=1567550022949710980,finish=1567550034399562342,duration=11449851362
```
</details>

**Comment by mvo5** [comment]
Both test failures (unreleated) are on 19.10 - interessting.

**Comment by mvo5** [comment]
This looks fine and the addition makes sense. Why not make architecture its own toplevel? This way its easier to parse in scripts grep|cut and we don't change existing output. Or am I missing something?

File: cmd/snap/cmd_version.go [hunk 1]
```diff
@@ -58,7 +58,11 @@ func printVersions(cli *client.Client) error {
 
 	fmt.Fprintf(w, "snap\t%s\n", cmd.Version)
 	fmt.Fprintf(w, "snapd\t%s\n", sv.Version)
-	fmt.Fprintf(w, "series\t%s\n", sv.Series)
+	if sv.Architecture != "" {
```
Comment (bboozzoo) [inline cmd/snap/cmd_version.go]: Why not
```
if sv.Architecture != "" {
    fmt.Fprintf(w, "architecture\t%s\n", sv.Architecture)
} else {
    fmt.Fprintf(w, "architecture\t-\n")
}
```

BTW. we can't use `arch` as Arch uses `ID=arch` in /etc/os-release and we print that out already:
```
snap    2.41-1
snapd   +git443.g7abba46
series  16
arch    -
kernel  5.2.11-arch1-1-ARCH
```

**Comment by chipaca** [comment]
@mvo @bboozzoo I played with a few options for the output:
as its own toplevel,
```
snap          133.7
snapd         2.41
series        16
ubuntu        16.04
kernel        4.4.0-159-generic
architecture  amd64
```
I didn't like how it pushed everything else so far to the right;
as part of the `kernel` line,
```
snap    133.7
snapd   2.41
series  16
ubuntu  16.04
kernel  4.4.0-159-generic on amd64
```

In the end I thought `series` was not ideal, but wasn't too terrible, and it didn't suffer from these other issues.

Of course, if we come up with a short header instead of `architecture` I'd love to use that instead.
```
snap    133.7
snapd   2.41
series  16
ubuntu  16.04
kernel  4.4.0-159-generic
on      amd64
```
...

**Comment by mvo5** [comment]
Thanks John for explaining this! Indeed, "architecture" is so long that it feels out of place a bit. Some ideas for "architecture": "machine", "cpu", "ISA", "processor" but none of this feels really great :/ Adding it to "series" looks actually more compelling than before :) (also as series is a bit useless right now as its 16 for some time). Sorry for being so indecisive.



P.S. Looking at this also makes me wonder (unrelated to this PR) why we did not use "key: value" here to look yamlish like in info. But lets not make this an even bigger (or brighter) bike-shed :-D

**Comment by bboozzoo** [comment]
Hmm
```
snap      2.41-1
snapd     +git443.g7abba46
series    16
arch      -
kernel    5.2.11-arch1-1-ARCH
...
host-arch amd64
host      amd64
on        amd64
```

OTOH, `arch   -` is confusing.

**Comment by pedronis** [comment]
one thing we could do is also capture the output of systemd-detect-virt (we use this in a couple of places already) and can be useful debug info as well:

then we could have:

if "none":
```
snap      2.41-1
...
host       amd64
```

if not "none"
```
snap     2.41-1
...
host     amd64 lxc
```

etc

**Comment by chipaca** [comment]
@pedronis this PR arbitrarily extends `/v2/system-info` to include the architecture `snapd` is running on; it could have just as easily checked the architecture `snap` is running on, as today those are the same thing unless the user is doing something particularly imaginative. The same argument can be made for virtualisation. So, I need to ask: does including them in snapd's system-info seem sane, and if so, should it also include virtualisation, and if so should that be a seaprate key or would it be rolled into architecture (now called host in the api perhaps?).

Personally I think, *if* it goes in the API (and I've got a slight preference for this), then it should be separate, i.e. `architecture` and `virtualisation`. But no really strong feelings either way.

**Comment by pedronis** [comment]
@chipaca yes including architecture and virtualization in the API makes sense, yes two separate fields, though under a "host" top field ?  but using a string to convey two values, no

File: cmd/snap/cmd_version.go [hunk 2]
```diff
@@ -68,6 +68,10 @@ func printVersions(cli *client.Client) error {
 	if sv.KernelVersion != "" {
 		fmt.Fprintf(w, "kernel\t%s\n", sv.KernelVersion)
 	}
+	if sv.Architecture != "" {
+		fmt.Fprintf(w, "host\t%s %s\n", sv.Architecture, sv.Virtualization)
```
[the remark this answers is not in the document]
Comment (chipaca) [reply cmd/snap/cmd_version.go]: we already do, it'll never give you `"none"`, just `""`

[the remark this answers is not in the document]
Comment (chipaca) [reply cmd/snap/cmd_version.go]: (because `systemd-detect-virt` returns error together with `"none"`)

**Pull request merged**